File: tiny/tiny-os/v6/fs.c
// Claude Code, Copyright (C) 2026 Yoann Padioleau, LGPL (see TinyC.ml)
//
// tiny-os v6: the file system (xv6's virtio_disk.c, bio.c, fs.c, less
// the log), on the format TinyMkfs.ml makes: a superblock, the inodes,
// a bitmap, the data. The layers, from the bottom: the disk's blocks;
// the buffer cache (a block in memory, one process at a time); the
// inodes (a file's blocks, its size); the directories (names to
// inodes); the paths. No log: a crash between two writes may leave the
// disk inconsistent, as Unix's was before fsck.
#include "defs.h"

struct superblock sb;

// the inodes in use, in memory: ref counts the pointers to one; busy is
// its lock (ilock); valid says its fields were read from the disk; gone
// that its name was unlinked, so that the last iput frees it
struct inode inodes[NINODE];
struct spinlock icache_lock;

// the disk, polled: tiny-machine moves the block at once, so there is
// nothing to sleep for (its interrupt is not enabled)
void
diskrw(struct buf *b, int write)
{
	*(uint*)DISK_BLOCK = b->blockno;
	*(uint*)DISK_ADDR = (uint)b->data;
	*(uint*)DISK_CMD = 1 + write;
	while(*(uint*)DISK_STATUS == 0)
		;
	*(uint*)DISK_STATUS = 0;
}

// ---------------------------------------------------------------- the buffer cache

struct buf bcache[NBUF];
struct spinlock bcache_lock;
uint bclock;

// the block's buffer, busy: this process's until brelse
struct buf*
bread(uint blockno)
{
	struct buf *b, *lru;

	acquire(&bcache_lock);
	for(;;){
		lru = 0;
		for(b = bcache; b < &bcache[NBUF]; b++){
			if(b->blockno == blockno && (b->valid || b->busy))
				break;
			if(!b->busy && (lru == 0 || b->used < lru->used))
				lru = b;
		}
		if(b < &bcache[NBUF]){
			if(b->busy){
				sleep(b, &bcache_lock);
				continue;
			}
			b->busy = 1;
			release(&bcache_lock);
			return b;
		}
		if(lru == 0)
			panic("bread: no buffers");
		lru->blockno = blockno;
		lru->valid = 0;
		lru->busy = 1;
		release(&bcache_lock);
		diskrw(lru, 0);
		lru->valid = 1;
		return lru;
	}
}

void
bwrite(struct buf *b)
{
	diskrw(b, 1);
}

void
brelse(struct buf *b)
{
	acquire(&bcache_lock);
	b->busy = 0;
	b->used = ++bclock;
	wakeup(b);
	release(&bcache_lock);
}

void
fsinit(void)
{
	struct buf *b;

	initlock(&bcache_lock, "bcache");
	initlock(&icache_lock, "icache");
	b = bread(0);
	memmove((char*)&sb, (char*)b->data, sizeof sb);
	brelse(b);
	if(sb.magic != FSMAGIC)
		panic("not a tiny-os v6 file system");
}

// ---------------------------------------------------------------- blocks

// a free block, zeroed, marked used in the bitmap
uint
balloc(void)
{
	struct buf *b, *z;
	uint n, mask;

	b = bread(sb.bmapstart);
	for(n = sb.datastart; n < sb.size; n++){
		mask = 1 << (n % 8);
		if((b->data[n / 8] & mask) == 0){
			b->data[n / 8] = b->data[n / 8] | mask;
			bwrite(b);
			brelse(b);
			z = bread(n);
			memset(z->data, 0, BSIZE);
			bwrite(z);
			brelse(z);
			return n;
		}
	}
	panic("balloc: out of blocks");
	return 0;
}

void
bfree(uint n)
{
	struct buf *b;

	b = bread(sb.bmapstart);
	b->data[n / 8] = b->data[n / 8] & ~(1 << (n % 8));
	bwrite(b);
	brelse(b);
}

// ---------------------------------------------------------------- inodes

struct inode*
iget(uint inum)
{
	struct inode *ip, *empty;

	acquire(&icache_lock);
	empty = 0;
	for(ip = inodes; ip < &inodes[NINODE]; ip++){
		if(ip->ref > 0 && ip->inum == inum){
			ip->ref++;
			release(&icache_lock);
			return ip;
		}
		if(empty == 0 && ip->ref == 0)
			empty = ip;
	}
	if(empty == 0)
		panic("iget: no inodes");
	empty->inum = inum;
	empty->ref = 1;
	empty->valid = 0;
	empty->busy = 0;
	empty->gone = 0;
	release(&icache_lock);
	return empty;
}

struct inode*
idup(struct inode *ip)
{
	acquire(&icache_lock);
	ip->ref++;
	release(&icache_lock);
	return ip;
}

uint
iblock(uint inum)
{
	return sb.inodestart + inum / (BSIZE / 64);
}

void
ilock(struct inode *ip)
{
	struct buf *b;
	struct dinode *d;

	acquire(&icache_lock);
	while(ip->busy)
		sleep(ip, &icache_lock);
	ip->busy = 1;
	release(&icache_lock);
	if(!ip->valid){
		b = bread(iblock(ip->inum));
		d = (struct dinode*)b->data + ip->inum % (BSIZE / 64);
		ip->type = d->type;
		ip->major = d->major;
		ip->size = d->size;
		memmove((char*)ip->addrs, (char*)d->addrs, sizeof ip->addrs);
		brelse(b);
		ip->valid = 1;
		if(ip->type == 0)
			panic("ilock: a free inode");
	}
}

void
iunlock(struct inode *ip)
{
	acquire(&icache_lock);
	ip->busy = 0;
	wakeup(ip);
	release(&icache_lock);
}

// the inode's fields to the disk
void
iupdate(struct inode *ip)
{
	struct buf *b;
	struct dinode *d;

	b = bread(iblock(ip->inum));
	d = (struct dinode*)b->data + ip->inum % (BSIZE / 64);
	d->type = ip->type;
	d->major = ip->major;
	d->size = ip->size;
	memmove((char*)d->addrs, (char*)ip->addrs, sizeof ip->addrs);
	bwrite(b);
	brelse(b);
}

// a free inode on the disk, of a type
struct inode*
ialloc(int type)
{
	uint inum;
	struct buf *b;
	struct dinode *d;

	for(inum = 1; inum < sb.ninodes; inum++){
		b = bread(iblock(inum));
		d = (struct dinode*)b->data + inum % (BSIZE / 64);
		if(d->type == 0){
			memset(d, 0, sizeof *d);
			d->type = type;
			bwrite(b);
			brelse(b);
			return iget(inum);
		}
		brelse(b);
	}
	panic("ialloc: no inodes");
	return 0;
}

// the file's blocks freed, its size 0 (ip locked)
void
itrunc(struct inode *ip)
{
	int i;
	struct buf *b;
	uint *a;

	for(i = 0; i < NDIRECT; i++)
		if(ip->addrs[i]){
			bfree(ip->addrs[i]);
			ip->addrs[i] = 0;
		}
	if(ip->addrs[NDIRECT]){
		b = bread(ip->addrs[NDIRECT]);
		a = (uint*)b->data;
		for(i = 0; i < NINDIRECT; i++)
			if(a[i])
				bfree(a[i]);
		brelse(b);
		bfree(ip->addrs[NDIRECT]);
		ip->addrs[NDIRECT] = 0;
	}
	ip->size = 0;
	iupdate(ip);
}

// a pointer dropped; the last, of an unlinked file, frees it
void
iput(struct inode *ip)
{
	acquire(&icache_lock);
	if(ip->ref == 1 && ip->valid && ip->gone){
		ip->busy = 1;
		release(&icache_lock);
		itrunc(ip);
		ip->type = 0;
		iupdate(ip);
		ip->valid = 0;
		acquire(&icache_lock);
		ip->busy = 0;
	}
	ip->ref--;
	release(&icache_lock);
}

void
iunlockput(struct inode *ip)
{
	iunlock(ip);
	iput(ip);
}

// the disk block of the file's n-th, allocated if need be
uint
bmap(struct inode *ip, uint n)
{
	struct buf *b;
	uint *a, addr;

	if(n < NDIRECT){
		if(ip->addrs[n] == 0)
			ip->addrs[n] = balloc();
		return ip->addrs[n];
	}
	n -= NDIRECT;
	if(n >= NINDIRECT)
		panic("bmap: too large");
	if(ip->addrs[NDIRECT] == 0)
		ip->addrs[NDIRECT] = balloc();
	b = bread(ip->addrs[NDIRECT]);
	a = (uint*)b->data;
	if((addr = a[n]) == 0){
		addr = a[n] = balloc();
		bwrite(b);
	}
	brelse(b);
	return addr;
}

void
stati(struct inode *ip, struct stat *st)
{
	st->type = ip->type;
	st->ino = ip->inum;
	st->size = ip->size;
}

// a process's memory (user) or the kernel's, as a copy's other end
int
either_copyout(int user, uint dst, char *src, uint n)
{
	if(user)
		return copyout(myproc()->pagetable, dst, src, n);
	memmove((char*)dst, src, n);
	return 0;
}

int
either_copyin(char *dst, int user, uint src, uint n)
{
	if(user)
		return copyin(myproc()->pagetable, dst, src, n);
	memmove(dst, (char*)src, n);
	return 0;
}

// n bytes at off, to dst; the count read (ip locked)
int
readi(struct inode *ip, int user, uint dst, uint off, uint n)
{
	uint tot, m;
	struct buf *b;

	if(off > ip->size)
		return 0;
	if(off + n > ip->size)
		n = ip->size - off;
	for(tot = 0; tot < n; tot += m, off += m, dst += m){
		b = bread(bmap(ip, off / BSIZE));
		m = BSIZE - off % BSIZE;
		if(m > n - tot)
			m = n - tot;
		if(either_copyout(user, dst, (char*)b->data + off % BSIZE, m) < 0){
			brelse(b);
			return -1;
		}
		brelse(b);
	}
	return n;
}

int
writei(struct inode *ip, int user, uint src, uint off, uint n)
{
	uint tot, m;
	struct buf *b;

	if(off > ip->size || off + n > (NDIRECT + NINDIRECT) * BSIZE)
		return -1;
	for(tot = 0; tot < n; tot += m, off += m, src += m){
		b = bread(bmap(ip, off / BSIZE));
		m = BSIZE - off % BSIZE;
		if(m > n - tot)
			m = n - tot;
		if(either_copyin((char*)b->data + off % BSIZE, user, src, m) < 0){
			brelse(b);
			break;
		}
		bwrite(b);
		brelse(b);
	}
	if(off > ip->size)
		ip->size = off;
	iupdate(ip);
	return tot;
}

// ---------------------------------------------------------------- directories and paths

// the entry named name in dp, its offset in *poff (dp locked)
struct inode*
dirlookup(struct inode *dp, char *name, uint *poff)
{
	uint off;
	struct dirent de;

	if(dp->type != T_DIR)
		panic("dirlookup");
	for(off = 0; off < dp->size; off += sizeof de){
		readi(dp, 0, (uint)&de, off, sizeof de);
		if(de.inum != 0 && strncmp(name, de.name, DIRSIZ) == 0){
			if(poff)
				*poff = off;
			return iget(de.inum);
		}
	}
	return 0;
}

// a new entry in dp (locked), in a free slot or at the end
int
dirlink(struct inode *dp, char *name, uint inum)
{
	uint off;
	struct dirent de;
	struct inode *ip;

	if((ip = dirlookup(dp, name, 0)) != 0){
		iput(ip);
		return -1;
	}
	for(off = 0; off < dp->size; off += sizeof de){
		readi(dp, 0, (uint)&de, off, sizeof de);
		if(de.inum == 0)
			break;
	}
	memset(&de, 0, sizeof de);
	de.inum = inum;
	memmove(de.name, name, strlen(name) < DIRSIZ ? strlen(name) : DIRSIZ);
	if(writei(dp, 0, (uint)&de, off, sizeof de) != sizeof de)
		return -1;
	return 0;
}

// the path's next element into name (DIRSIZ at most), what follows it
char*
skipelem(char *path, char *name)
{
	char *s;
	int len;

	while(*path == '/')
		path++;
	if(*path == 0)
		return 0;
	s = path;
	while(*path != '/' && *path != 0)
		path++;
	len = path - s;
	if(len > DIRSIZ)
		len = DIRSIZ;
	memmove(name, s, len);
	if(len < DIRSIZ)
		name[len] = 0;
	while(*path == '/')
		path++;
	return path;
}

// the path's inode, or its parent's (the last element in name)
struct inode*
namex(char *path, int parent, char *name)
{
	struct inode *ip, *next;

	if(*path == '/')
		ip = iget(ROOTINO);
	else
		ip = idup(myproc()->cwd);
	while((path = skipelem(path, name)) != 0){
		ilock(ip);
		if(ip->type != T_DIR){
			iunlockput(ip);
			return 0;
		}
		if(parent && *path == 0){
			iunlock(ip);
			return ip;
		}
		if((next = dirlookup(ip, name, 0)) == 0){
			iunlockput(ip);
			return 0;
		}
		iunlockput(ip);
		ip = next;
	}
	if(parent){
		iput(ip);
		return 0;
	}
	return ip;
}

struct inode*
namei(char *path)
{
	char name[DIRSIZ + 1];

	return namex(path, 0, name);
}

struct inode*
nameiparent(char *path, char *name)
{
	return namex(path, 1, name);
}

File: kernel/xv6/Fs.ml
(* Claude Code
 *
 * Copyright (C) 2026 Yoann Padioleau
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public License
 * (LGPL) as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 *)
(* See Fs.mli *)

open Types

module Phys = Machine.Phys

(*****************************************************************************)
(* The disk *)
(*****************************************************************************)

let base = Machine.fs_base ()
let magic = 0x10203040

(* the block size, the disk's: where block 1, the superblock, starts
 * (block 0 is empty). xv6's ports differ (arm-pi1's 512, arm64-pi4's
 * 1024: each fork's driver's), their format otherwise the same *)
let bsize =
  if Phys.get32 (base + 512) = magic then 512
  else if Phys.get32 (base + 1024) = magic then 1024
  else Machine.panic "fs: not an xv6 file system"

(* fs.h, param.h *)
let ndirect = 58
let nindirect = bsize / 4
let maxfile = ndirect + nindirect
let ipb = bsize / 256                       (* inodes per block: a dinode is 256 bytes *)
let dirsiz = 14
let rootino = 1
let ninode = 50

let block b = base + (b * bsize)

(* the superblock, block 1: magic, size, nblocks, ninodes, nlog,
 * logstart, inodestart, bmapstart *)
let sb k = Phys.get32 (block 1 + (4 * k))
let fsize = sb 1
let ninodes = sb 3
let inodestart = sb 6
let bmapstart = sb 7

(*****************************************************************************)
(* An inode's fields, on the disk *)
(*****************************************************************************)

(* a dinode: short type, major, minor, nlink (ownerid, groupid, mode:
 * amd64-jserv's), uint size at 16, addrs at 20 *)
type field = Half of int | Word of int

let i_type = Half 0
let i_major = Half 2
let i_minor = Half 4
let i_nlink = Half 6
let i_size = Word 16
let i_addr k = Word (20 + (4 * k))          (* k = ndirect: the indirect block *)

let dinode inum = block ((inum / ipb) + inodestart) + ((inum mod ipb) * 256)

let get ip f =
  match f with
  | Half o -> Phys.get16 (dinode ip.inum + o)
  | Word o -> Phys.get32 (dinode ip.inum + o)

let set ip f v =
  match f with
  | Half o -> Phys.set16 (dinode ip.inum + o) v
  | Word o -> Phys.set32 (dinode ip.inum + o) v

let itype ip = match get ip i_type with 1 -> Dir | 2 -> File | 3 -> Devnode | _ -> Free
let set_itype ip t = set ip i_type (match t with Free -> 0 | Dir -> 1 | File -> 2 | Devnode -> 3)

(*****************************************************************************)
(* Blocks *)
(*****************************************************************************)

(* a block's bit in the bitmap: its byte's address, the mask *)
let bit b = block ((b / (bsize * 8)) + bmapstart) + ((b mod (bsize * 8)) / 8), 1 lsl (b mod 8)

(* the first free block, marked used, zeroed *)
let balloc () =
  let rec go b =
    if b >= fsize then Machine.panic "balloc: out of blocks"
    else
      let a, m = bit b in
      if Phys.get8 a land m = 0 then begin
        Phys.set8 a (Phys.get8 a lor m);
        Phys.zero (block b) bsize;
        b
      end
      else go (b + 1) in
  go 0

let bfree b =
  let a, m = bit b in
  if Phys.get8 a land m = 0 then Machine.panic "freeing free block";
  Phys.set8 a (Phys.get8 a land lnot m)

(* the disk block holding the file's block [bn], allocated if none yet *)
let bmap ip bn =
  let ensure read write = match read () with 0 -> let b = balloc () in write b; b | b -> b in
  if bn < ndirect then ensure (fun () -> get ip (i_addr bn)) (set ip (i_addr bn))
  else if bn < maxfile then begin
    let ind = ensure (fun () -> get ip (i_addr ndirect)) (set ip (i_addr ndirect)) in
    let e = block ind + (4 * (bn - ndirect)) in
    ensure (fun () -> Phys.get32 e) (Phys.set32 e)
  end
  else Machine.panic "bmap: out of range"

(* the file's blocks all freed, its size 0 *)
let itrunc ip =
  for k = 0 to ndirect - 1 do
    let b = get ip (i_addr k) in
    if b <> 0 then begin bfree b; set ip (i_addr k) 0 end
  done;
  let ind = get ip (i_addr ndirect) in
  if ind <> 0 then begin
    for j = 0 to nindirect - 1 do
      let b = Phys.get32 (block ind + (4 * j)) in
      if b <> 0 then bfree b
    done;
    bfree ind;
    set ip (i_addr ndirect) 0
  end;
  set ip i_size 0

(*****************************************************************************)
(* Inodes in use *)
(*****************************************************************************)

(* xv6's icache, its slots the inodes held: NINODE at most *)
let icache : inode list ref = ref []

let iget inum =
  match List.filter (fun ip -> ip.inum = inum) !icache with
  | ip :: _ -> ip.iref <- ip.iref + 1; ip
  | [] ->
      if List.length !icache >= ninode then Machine.panic "iget: no inodes";
      let ip = { inum = inum; iref = 1 } in
      icache := ip :: !icache;
      ip

let idup ip = ip.iref <- ip.iref + 1; ip

(* the last reference to an inode no directory names: the inode freed *)
let iput ip =
  if ip.iref = 1 && get ip i_nlink = 0 then begin
    itrunc ip;
    set_itype ip Free
  end;
  ip.iref <- ip.iref - 1;
  if ip.iref = 0 then icache := List.filter (fun i -> i != ip) !icache

(* a free inode made [t] *)
let ialloc t =
  let rec go inum =
    if inum >= ninodes then Machine.panic "ialloc: no inodes"
    else if itype { inum = inum; iref = 0 } = Free then begin
      Phys.zero (dinode inum) 256;
      let ip = iget inum in
      set_itype ip t;
      ip
    end
    else go (inum + 1) in
  go 1

(*****************************************************************************)
(* An inode's bytes *)
(*****************************************************************************)

(* [n] bytes at [off], fewer at the end, none past it (xv6-riscv's
 * readi: 0, not -1) *)
let readi ip off n =
  let size = get ip i_size in
  if off > size then Some ""
  else if off < 0 || n < 0 then None
  else begin
    let n = min n (size - off) in
    let b = Buffer.create n in
    let rec go off k =
      if k > 0 then begin
        let m = min k (bsize - (off mod bsize)) in
        Buffer.add_string b (Phys.read (block (bmap ip (off / bsize)) + (off mod bsize)) m);
        go (off + m) (k - m)
      end in
    go off n;
    Some (Buffer.contents b)
  end

(* [s] written at [off], the file grown: its length, or -1 (a hole, or
 * past MAXFILE) *)
let writei ip off s =
  let n = String.length s in
  if off < 0 || off > get ip i_size || off + n > maxfile * bsize then -1
  else begin
    let rec go off i =
      if i < n then begin
        let m = min (n - i) (bsize - (off mod bsize)) in
        Phys.write (block (bmap ip (off / bsize)) + (off mod bsize)) (String.sub s i m);
        go (off + m) (i + m)
      end in
    go off 0;
    if n > 0 && off + n > get ip i_size then set ip i_size (off + n);
    n
  end

(* xv6-riscv's readi to the user (a read(2)): [n] bytes at [off], a
 * block's piece at a time to [dst] (the user's buffer, by offset): the
 * bytes read, or -1 when [dst] cannot take one (the pieces before it
 * copied). [n] a C uint: a negative count wraps, past the end (0) or to
 * the whole file *)
let readi_to ip off n dst =
  let size = get ip i_size in
  if off > size || (n < 0 && off + n >= 0) then 0
  else begin
    let n = if n < 0 then size - off else min n (size - off) in
    let rec go tot off =
      if tot >= n then tot
      else begin
        let m = min (n - tot) (bsize - (off mod bsize)) in
        if dst tot (Phys.read (block (bmap ip (off / bsize)) + (off mod bsize)) m) then go (tot + m) (off + m) else -1
      end in
    go 0 off
  end

(* xv6-riscv's writei from the user: [n] bytes from [src] at [off], a
 * block's piece at a time, stopping at one [src] cannot give; the bytes
 * written, the file grown to them; -1 past the end, past MAXFILE, or a
 * negative (a C uint: huge) count *)
let writei_from ip off n src =
  if n < 0 || off > get ip i_size || off + n > maxfile * bsize then -1
  else begin
    let rec go tot off =
      if tot >= n then tot
      else begin
        let m = min (n - tot) (bsize - (off mod bsize)) in
        match src tot m with
        | None -> tot
        | Some s -> Phys.write (block (bmap ip (off / bsize)) + (off mod bsize)) s; go (tot + m) (off + m)
      end in
    let tot = go 0 off in
    if off + tot > get ip i_size then set ip i_size (off + tot);
    tot
  end

(* struct stat: int dev, uint ino, short type, short nlink, then at 16
 * uint64 size; its padding (12-15) the user's, untouched *)
let stat_head ip =
  Machine.le32 1 ^ Machine.le32 ip.inum ^ Machine.le16 (get ip i_type) ^ Machine.le16 (get ip i_nlink)
let stat_size ip = Machine.le32 (get ip i_size) ^ Machine.le32 0

(*****************************************************************************)
(* Directories *)
(*****************************************************************************)

(* a dirent: ushort inum, 14 bytes of name (NUL-padded, or not ended) *)
let dirent_inum s o = Char.code s.[o] lor (Char.code s.[o + 1] lsl 8)
let dirent_name s o =
  let rec len k = if k < dirsiz && s.[o + 2 + k] <> '\000' then len (k + 1) else k in
  String.sub s (o + 2) (len 0)

let entries dp = match readi dp 0 (get dp i_size) with Some s -> s | None -> ""

(* [name]'s inode (held), and its entry's offset *)
let dirlookup dp name =
  let s = entries dp in
  let rec go off =
    if off + 16 > String.length s then None
    else if dirent_inum s off <> 0 && dirent_name s off = name then Some (iget (dirent_inum s off), off)
    else go (off + 16) in
  go 0

(* a new entry, in the first free one or at the end; -1 if [name] is
 * there *)
let dirlink dp name inum =
  match dirlookup dp name with
  | Some (ip, _) -> iput ip; -1
  | None ->
      let s = entries dp in
      let rec free off = if off + 16 > String.length s || dirent_inum s off = 0 then off else free (off + 16) in
      let padded = name ^ String.make (dirsiz - String.length name) '\000' in
      if writei dp (free 0) (Machine.le16 inum ^ padded) <> 16 then Machine.panic "dirlink";
      0

(* only "." and ".." *)
let isdirempty dp =
  let s = entries dp in
  let rec go off = off + 16 > String.length s || (dirent_inum s off = 0 && go (off + 16)) in
  go 32

(*****************************************************************************)
(* Paths *)
(*****************************************************************************)

(* the element at [i] (its first 14 bytes: DIRSIZ) and where the next
 * starts, the slashes skipped; None at the end *)
let skipelem path i =
  let n = String.length path in
  let rec skip i = if i < n && path.[i] = '/' then skip (i + 1) else i in
  let rec stop j = if j < n && path.[j] <> '/' then stop (j + 1) else j in
  let i = skip i in
  if i = n then None
  else
    let j = stop i in
    Some (String.sub path i (min (j - i) dirsiz), skip j)

(* the path's inode, or ([parent]) its directory's and its last name:
 * held; from the root or the process's directory *)
let namex path parent =
  let start = if String.length path > 0 && path.[0] = '/' then iget rootino else idup (Proc.myproc ()).cwd in
  let rec go ip i =
    match skipelem path i with
    | None -> if parent then begin iput ip; None end else Some (ip, "")
    | Some (name, j) ->
        if itype ip <> Dir then begin iput ip; None end
        else if parent && j = String.length path then Some (ip, name)
        else match dirlookup ip name with
          | None -> iput ip; None
          | Some (next, _) -> iput ip; go next j in
  go start 0

let namei path = match namex path false with Some (ip, _) -> Some ip | None -> None
let nameiparent path = namex path true

(*****************************************************************************)
(* Operations on names (sysfile.c's) *)
(*****************************************************************************)

(* the inode at [path], made [t] if new (held); an existing file does for
 * a file, nothing else does *)
let create path t major minor =
  match nameiparent path with
  | None -> None
  | Some (dp, name) ->
      match dirlookup dp name with
      | Some (ip, _) ->
          iput dp;
          if t = File && (itype ip = File || itype ip = Devnode) then Some ip else begin iput ip; None end
      | None ->
          let ip = ialloc t in
          set ip i_major major;
          set ip i_minor minor;
          set ip i_nlink 1;
          if t = Dir then begin
            (* ".." names the parent; "." does not count (no cycle) *)
            set dp i_nlink (get dp i_nlink + 1);
            if dirlink ip "." ip.inum < 0 || dirlink ip ".." dp.inum < 0 then Machine.panic "create dots"
          end;
          if dirlink dp name ip.inum < 0 then Machine.panic "create: dirlink";
          iput dp;
          Some ip

let link old new_ =
  match namei old with
  | None -> -1
  | Some ip when itype ip = Dir -> iput ip; -1
  | Some ip ->
      set ip i_nlink (get ip i_nlink + 1);
      let ok = match nameiparent new_ with
        | None -> false
        | Some (dp, name) -> let r = dirlink dp name ip.inum in iput dp; r = 0 in
      if not ok then set ip i_nlink (get ip i_nlink - 1);
      iput ip;
      if ok then 0 else -1

let unlink path =
  match nameiparent path with
  | None -> -1
  | Some (dp, name) ->
      let r =
        if name = "." || name = ".." then -1
        else match dirlookup dp name with
          | None -> -1
          | Some (ip, off) ->
              if get ip i_nlink < 1 then Machine.panic "unlink: nlink < 1";
              if itype ip = Dir && not (isdirempty ip) then begin iput ip; -1 end
              else begin
                if writei dp off (String.make 16 '\000') <> 16 then Machine.panic "unlink: writei";
                if itype ip = Dir then set dp i_nlink (get dp i_nlink - 1);
                set ip i_nlink (get ip i_nlink - 1);
                iput ip;
                0
              end in
      iput dp;
      r
